ModuleThreeFirstTaskConsole/FileSystemVisitor.cs

```diff
@@ -0,0 +1,229 @@
+        public event EventHandler<FileSystemEventArgs> FilteredFileFound;
+
+        /// <summary>
+        /// Events that happen when filtered directory is found.
+        /// </summary>
+        public event EventHandler<FileSystemEventArgs> FilteredDirectoryFound;
+
+        /// <summary>
+        /// State of searching. If it's true on the next step except for search start the visitor will not proceed to work.
+        /// </summary>
+        public bool Stoped { get; set; }
+
+        /// <summary>
+        /// Search files in provided fileSystem.
+        /// </summary>
+        /// <returns>Files from directories.</returns>
+        public async IAsyncEnumerable<string> Search()
+        {
+            OnSearchStarted(fileSystem);
+            await foreach (var info in GetAllFromCurrentDirectory(fileSystem))
+            {
+                if (Stoped)
+                {
+                    break;
+                }
+
+                yield return GetNameWithoutLongPath(info);
+            }
+
+            OnSearchEnded(fileSystem);
+        }
+
+        /// <summary>
+        /// Invoke methods from SearchStarted event.
+        /// </summary>
+        /// <param name="info">Provided info.</param>
+        protected virtual void OnSearchStarted(FileSystemInfo info)
+        {
+            Stoped = false;
+            SearchStarted?.Invoke(this, info);
+        }
+
+        /// <summary>
+        /// Invoke methods from SearchEnded event.
+        /// </summary>
+        /// <param name="info">Provided info.</param>
+        protected virtual void OnSearchEnded(FileSystemInfo info)
+        {
+            Stoped = true;
+            SearchEnded?.Invoke(this, info);
+        }
+
+        /// <summary>
+        /// Invoke methods from FileFound event.
+        /// </summary>
+        /// <param name="args"></param>
+        protected virtual void OnFileFound(FileSystemEventArgs args)
+        {
+            FileFound?.Invoke(this, args);
+            Stoped = args.Stop ? args.Stop : Stoped;
+        }
+
+        /// <summary>
+        /// Invoke methods from DirectoryFound event.
+        /// </summary>
+        /// <param name="args"></param>
+        protected virtual void OnDirectoryFound(FileSystemEventArgs args)
+        {
+            DirectoryFound?.Invoke(this, args);
+            Stoped = args.Stop ? args.Stop : Stoped;
+        }
+
+        /// <summary>
+        /// Invoke methods from FilteredFileFound event.
+        /// </summary>
+        /// <param name="args"></param>
+        protected virtual void OnFilteredFileFound(FileSystemEventArgs args)
+        {
+            FilteredFileFound?.Invoke(this, args);
+            Stoped = args.Stop ? args.Stop : Stoped;
+        }
+
+        /// <summary>
+        /// Invoke methods from FilteredDirectoryFound event.
+        /// </summary>
+        /// <param name="args"></param>
+        protected virtual void OnFilteredDirectoryFound(FileSystemEventArgs args)
```

Comment:
The same method as for `OnFilteredFileFound` and others with type `(FileSystemEventArgs args) => void`. All of them can be refactored into one.

Reply:
Fixed in cea260a